### PR TITLE
feat(trainer): Support Framework Labels in Runtimes

### PR DESCRIPTION
**Pull request by andreyvelich**
Fixes: https://github.com/kubeflow/sdk/pull/31, https://github.com/kubeflow/sdk/issues/29

I refactored SDK to support runtime labels and give users ability to override the default runtime images.

/assign @Electronic-Waste @astefanutti @kramaranya @jskswamy @ned2 @eoinfennessy @briangallagher

### CODE REVIEW

File: python/kubeflow/trainer/types/types.py [hunk 1]
```diff
@@ -149,26 +149,32 @@ class BuiltinTrainer:
...
-    entrypoint: Optional[List[str]] = None
     accelerator_count: Union[str, float, int] = constants.UNKNOWN
+    __command: tuple[str, ...] = field(init=False, repr=False)
+
+    @property
+    def command(self) -> tuple[str, ...]:
+        return self.__command
+
+    def set_command(self, command: tuple[str, ...]):
+        self.__command = command
```
Comment (andreyvelich) [inline python/kubeflow/trainer/types/types.py:179]: I make this __command private, so it should not be accessed or modified by users.

File: python/kubeflow/trainer/constants/constants.py [hunk 2]
```diff
@@ -120,27 +124,49 @@
...
+# The exec script to embed training function into container command.
+# __ENTRYPOINT__ depends on the MLPolicy, func_code and func_file is substituted in the `train` API.
+EXEC_FUNC_SCRIPT = textwrap.dedent(
+    """
+        read -r -d '' SCRIPT << EOM\n
+        {func_code}
+        EOM
+        printf "%s" \"$SCRIPT\" > \"{func_file}\"
+        __ENTRYPOINT__ \"{func_file}\""""
+)
```
Comment (andreyvelich) [inline python/kubeflow/trainer/constants/constants.py:136]: For runtimes with CustomTrainer, I insert the complete `command` when Runtime object is created: 

https://github.com/kubeflow/sdk/blob/f3e90202b9af447afd30e03fcace9f4b842c16cb/python/kubeflow/trainer/utils/utils.py#L155-L164

I think, this is much clear approach since we always have the final  container `command` in `runtime.trainer.__command`.

File: python/kubeflow/trainer/types/types.py [hunk 3]
```diff
@@ -229,57 +235,3 @@ class Initializer:
 
     dataset: Optional[HuggingFaceDatasetInitializer] = None
     model: Optional[HuggingFaceModelInitializer] = None
-
-
-# The dict where key is the container image and value its representation.
-# Each Trainer representation defines trainer parameters (e.g. type, framework, entrypoint).
-# TODO (andreyvelich): We should allow user to overrides the default image names.
-ALL_TRAINERS: Dict[str, RuntimeTrainer] = {
```
Comment (astefanutti) [inline python/kubeflow/trainer/types/types.py:239]: Would keeping the existing mapping, but with the framework as key instead of the image, make things tidier / more structured rather than replacing that with individual strings?

File: python/kubeflow/trainer/constants/constants.py [hunk 4]
```diff
@@ -53,6 +54,9 @@
 # For example, what PodTemplate must be overridden by TrainJob's .spec.trainer APIs.
 TRAINJOB_ANCESTOR_LABEL = "trainer.kubeflow.org/trainjob-ancestor-step"
 
+# The label key to identify ML framework that runtime uses (e.g. torch, deepspeed, torchtune, etc.)
+RUNTIME_FRAMEWORK_LABEL = "trainer.kubeflow.org/framework"
+
```
Comment (Electronic-Waste) [inline python/kubeflow/trainer/constants/constants.py:59]: Shall we also define `TRAINER_TYPE_LABEL`?

File: python/kubeflow/trainer/types/types.py [hunk 5]
```diff
@@ -39,7 +39,7 @@ class CustomTrainer:
 
     func: Callable
     func_args: Optional[Dict] = None
-    packages_to_install: Optional[List[str]] = None
+    packages_to_install: Optional[list[str]] = None
```
Comment (Electronic-Waste) [inline python/kubeflow/trainer/types/types.py:43]: Why should we change from `List` to `list`?

File: python/kubeflow/trainer/types/types.py (hunk 3 above)
Comment (andreyvelich) [reply python/kubeflow/trainer/types/types.py:239]: But we don't need it, do we ? Theoretically, users should be able to use any arbitrary framework they define in the runtime labels with CustomTrainer, even that we don't support upstream (e.g. scikit-learn, tf, etc.).
The important mapping is only for BuiltinTrainers, since we need to map the `framework` name and BuiltinTrainer `configs`: https://github.com/kubeflow/sdk/blob/f3e90202b9af447afd30e03fcace9f4b842c16cb/python/kubeflow/trainer/types/types.py#L154.

If you want, we can create mapping for BuiltinTrainer configs, but for now we have only 1 config (e.g. TorchTune).

File: python/kubeflow/trainer/types/types.py [hunk 6]
```diff
@@ -149,26 +149,32 @@ class BuiltinTrainer:
     config: TorchTuneConfig
 
 
+# Change it to list: BUILTIN_CONFIGS, once we support more Builtin Trainer configs.
+TORCH_TUNE = (
```
Comment (Electronic-Waste) [inline python/kubeflow/trainer/types/types.py:155]: Do you prefer `TORCHTUNE` or `TORCH_TUNE`?

File: python/kubeflow/trainer/api/trainer_client.py [hunk 7]
```diff
@@ -190,13 +194,13 @@ def train(
             # If users choose to use a custom training function.
             if isinstance(trainer, types.CustomTrainer):
                 trainer_crd = utils.get_trainer_crd_from_custom_trainer(
-                    trainer, runtime
+                    runtime, trainer
```
Comment (Electronic-Waste) [inline python/kubeflow/trainer/api/trainer_client.py]: Can you add validation here to ensure that users reference the right runtime, as we discussed here: https://cloud-native.slack.com/archives/C0742LDFZ4K/p1753710956860929

File: python/kubeflow/trainer/api/trainer_client.py [hunk 8]
```diff
@@ -190,13 +194,13 @@ def train(
...
                 )
 
             # If users choose to use a builtin trainer for post-training.
             elif isinstance(trainer, types.BuiltinTrainer):
                 trainer_crd = utils.get_trainer_crd_from_builtin_trainer(
-                    trainer, initializer
+                    runtime, trainer, initializer
```
Comment (Electronic-Waste) [inline python/kubeflow/trainer/api/trainer_client.py]: Same as above

File: python/kubeflow/trainer/types/types.py (hunk 3 above)
Comment (astefanutti) [reply python/kubeflow/trainer/types/types.py:239]: OK that makes sense. It's more an internal implementation detail anyway.

File: python/kubeflow/trainer/types/types.py (hunk 5 above)
Comment (andreyvelich) [reply python/kubeflow/trainer/types/types.py:43]: Python 3.9 introduce builtin support for those types: https://docs.python.org/3/whatsnew/3.9.html#type-hinting-generics-in-standard-collections

File: python/kubeflow/trainer/types/types.py (hunk 3 above)
Comment (Electronic-Waste) [reply python/kubeflow/trainer/types/types.py:239]: We should verify that:

1. CustomTrainer does not reference runtimes corresponding to BuiltinTrainer
2. BuiltTrainer only reference corresponding runtimes

File: python/kubeflow/trainer/constants/constants.py (hunk 4 above)
Comment (andreyvelich) [reply python/kubeflow/trainer/constants/constants.py:59]: No, we don't really need it as we discussed here: https://github.com/kubeflow/trainer/pull/2761#discussion_r2244636856

File: python/kubeflow/trainer/types/types.py (hunk 6 above)
Comment (andreyvelich) [reply python/kubeflow/trainer/types/types.py:155]: Maybe TORCH_TUNE is more accurate since we use camel case here: TorchTuneConfig.

File: python/kubeflow/trainer/api/trainer_client.py (hunk 8 above)
Comment (andreyvelich) [reply python/kubeflow/trainer/api/trainer_client.py]: Sure, thanks for reminder!

**Review by andreyvelich — COMMENTED** [review]
/approve
